Documentation/guides/messages/README.md

```diff
@@ -41,20 +41,24 @@
 + [XA0113](xa0113.md): Google Play requires that new applications must use a `$(TargetFrameworkVersion)` of v8.0 (API level 26) or above.
 + [XA0114](xa0114.md): Google Play requires that application updates must use a `$(TargetFrameworkVersion)` of v8.0 (API level 26) or above.
 + [XA0115](xa0115.md): Invalid value 'armeabi' in $(AndroidSupportedAbis). This ABI is no longer supported. Please update your project properties.
++ [XA0118](xa0118.md): Could not parse '{TargetMoniker}'
 
 ### XA1xxx Project Related
 
 + [XA1000](xa1000.md): There was an problem parsing {file}. This is likely due to incomplete or invalid xml.
 + [XA1001](xa1001.md): AndroidResgen: Warning while updating Resource XML '{filename}': {Message}
 + [XA1002](xa1002.md): We found a matching key '{Key}' for '{Item}'. But the casing was incorrect. Please correct the casing
 + [XA1003](xa1003.md): '{zip}' does not exist. Please rebuild the project.
++ [XA1004](xa1004.md): Attempting naive type name fixup for element with ID '{id}' and type '{managedType}'
 
 ### XA2xxx Linker
 
 ### XA3xxx AOT
 
 ### XA4xxx Code Generation
 
++ [XA4214](xa4214.md): Duplicate managed type found! Mappings between managed types and Java types must be unique. First Type: '{0}'; Second Type: '{1}'.
++ [XA4215](xa4215.md): Duplicate Java type found! Mappings between managed types and Java types must be unique. First Type: '{0}'; Second Type: '{1}'.
```

Comment:
@brendanzagaeski does the wording on one of these messages need to be flipped?  eg. 

`Mappings between managed types and Java types must be unique`
`Mappings between Java types and managed types must be unique`

I suspect the two checks are from the opposite point of view. So we should reword the message to reflect which point of view we were using? maybe?

Reply:
Yeah, ensuring clear language in the actual message that comes from the warnings is good to do. For this pass, I was only planning to update messages "opportunistically" if I saw cases where an idea came to mind immediately as I was reading them for a better wording. But I'd be happy to spend a little more time reviewing the wording as I go, if that'd be useful. For these 2 particular warnings, I also had a hunch that users might not run into them often (we'll know better once we start getting the telemetry from the warning codes). So I didn't spend too much time thinking about the wording yet.

Reply:
I took another look at this warning & error pair today. It occurred to me that it might be best to rephrase these messages to be more like the error that `Csc` gives for similar type conflicts. That led me to see that the current messages use a slightly different way of splitting up the type and assembly name compared to the `Csc` message, so getting everything to match up closely would take some extra work. Long story short, I took [a try at updating the messages](https://gist.github.com/brendanzagaeski/ca1a5092e92bdeee71843786eea14add), but I'm not sure if the new wording better overall. If it does look like an improvement, I can add it to this pull request.

Reply:
I think improving the wording is a good idea.  I'm unsure whether the current suggestion is a net improvement.

Consider a solution with two library projects, referenced by the app project, which contain the following class:

```csharp
using System;
using Android.Runtime;

namespace Lib1
{
  [Register ("examplelib.EmptyClass")]
  public class EmptyClass : Java.Lang.Object
  {
  }
}
```

This elicits the warning (newlines added for readability):

```
warning : Duplicate managed type found! Mappings between managed types and Java types must be unique.
 First Type: 'Lib1.EmptyClass, Lib1, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null';
Second Type: 'Lib1.EmptyClass, Lib2, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null'.
```

*Hilariously*, if we update the project structure so that Lib1 has `Lib1.EmptyClass` and Lib2 has `Lib2.EmptyClass` -- instead of *both* Lib1 and Lib2 providing `Lib1.EmptyClass` -- *then* we get an *error* (newlines added for readability):

```
error : Duplicate Java type found! Mappings between managed types and Java types must be unique.
 First Type: 'Lib1.EmptyClass, Lib1, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null';
Second Type: 'Lib2.EmptyClass, Lib2, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null'
```

I find this *hilarious*, as it implies one form of ambiguity is worse than another, even though either way we have ambiguity. :-/

That setup aside, if we imagine the suggested XA4214 warning message, we'd have:

```
The managed type 'Lib1.EmptyClass' exists as both 'Lib1.EmptyClass, Lib1, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null' and Lib1.EmptyClass, Lib2, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null'. An Android Callable Wrapper will only be generated for 'Lib1.EmptyClass, Lib1, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null'.
```

Is this better?  Not immediately, no; it's not actionable.  It doesn't inform why this might be a problem -- which arguably could/should instead be in the docs -- nor does it have any suggestion of how to fix it.

(The fact that we have assembly-qualified names in here also makes me think we'd be better off with multi-line messages here!)

Reply:
After [pondering the corresponding documentation](
https://github.com/xamarin/xamarin-android/pull/2258/files#r222787663), how's this for an improved warning message?

```
The managed type `Lib1.EmptyClass` exists in the assemblies `Lib1`, `Lib2`.  Please refactor the managed type names so that they are not identical.
```

C#-wise, this would require additional changes to `<GenerateJavaStubs/>` so that we can properly capture *all* the assemblies that contain duplicate types, instead of the current code which emits the warning as soon as it finds a duplicate, and then will re-issue the "same" warning every time.

For example, my previous examples have all had two assemblies (Lib1 & Lib2) with a `Lib1.EmptyClass` type.  What happens when there are *three* assemblies with the same type?  Four assemblies?  10?  With the current approach, we'll emit `assemblies.Length-1` warnings, which is potentially very noise and not as actionable.